modules/caddyhttp/server_test.go

```diff
@@ -144,3 +145,207 @@ func BenchmarkServer_LogRequest_WithTraceID(b *testing.B) {
+
+	server := &Server{
+		trustedProxies: &StaticIPRange{
+			ranges: []netip.Prefix{loopbackPrefix},
+		},
+		ClientIPHeaders: []string{"X-Forwarded-For"},
+	}
+
+	req := httptest.NewRequest("GET", "/", nil)
+	req.RemoteAddr = "127.0.0.1:12345"
+	req.Header.Set("X-Forwarded-For", "31.40.0.10")
+
+	trusted, clientIP := determineTrustedProxy(req, server)
+
+	assert.True(t, trusted)
+	assert.Equal(t, clientIP, "31.40.0.10")
+}
+
+func TestServer_DetermineTrustedProxy_UntrustedPrefix(t *testing.T) {
+	loopbackPrefix, _ := netip.ParsePrefix("127.0.0.1/8")
+
+	server := &Server{
+		trustedProxies: &StaticIPRange{
+			ranges: []netip.Prefix{loopbackPrefix},
+		},
+		ClientIPHeaders: []string{"X-Forwarded-For"},
+	}
+
+	req := httptest.NewRequest("GET", "/", nil)
+	req.RemoteAddr = "10.0.0.1:12345"
+	req.Header.Set("X-Forwarded-For", "31.40.0.10")
+
+	trusted, clientIP := determineTrustedProxy(req, server)
+
+	assert.False(t, trusted)
+	assert.Equal(t, clientIP, "10.0.0.1")
+}
+
+func TestServer_DetermineTrustedProxy_MultipleTrustedPrefixes(t *testing.T) {
+	loopbackPrefix, _ := netip.ParsePrefix("127.0.0.1/8")
+	localPrivatePrefix, _ := netip.ParsePrefix("10.0.0.0/8")
+
+	server := &Server{
+		trustedProxies: &StaticIPRange{
+			ranges: []netip.Prefix{loopbackPrefix, localPrivatePrefix},
+		},
+		ClientIPHeaders: []string{"X-Forwarded-For"},
+	}
+
+	req := httptest.NewRequest("GET", "/", nil)
+	req.RemoteAddr = "10.0.0.1:12345"
+	req.Header.Set("X-Forwarded-For", "31.40.0.10")
+
+	trusted, clientIP := determineTrustedProxy(req, server)
+
+	assert.True(t, trusted)
+	assert.Equal(t, clientIP, "31.40.0.10")
+}
+
+func TestServer_DetermineTrustedProxy_MultipleTrustedClientHeaders(t *testing.T) {
+	loopbackPrefix, _ := netip.ParsePrefix("127.0.0.1/8")
+	localPrivatePrefix, _ := netip.ParsePrefix("10.0.0.0/8")
+
+	server := &Server{
+		trustedProxies: &StaticIPRange{
+			ranges: []netip.Prefix{loopbackPrefix, localPrivatePrefix},
+		},
+		ClientIPHeaders: []string{"CF-Connecting-IP", "X-Forwarded-For"},
+	}
+
+	req := httptest.NewRequest("GET", "/", nil)
+	req.RemoteAddr = "10.0.0.1:12345"
+	req.Header.Set("CF-Connecting-IP", "1.1.1.1, 2.2.2.2")
+	req.Header.Set("X-Forwarded-For", "3.3.3.3, 4.4.4.4")
+
+	trusted, clientIP := determineTrustedProxy(req, server)
+
+	assert.True(t, trusted)
+	assert.Equal(t, clientIP, "1.1.1.1")
+}
+
+func TestServer_DetermineTrustedProxy_SkipTrustedPrivateHops(t *testing.T) {
+	localPrivatePrefix, _ := netip.ParsePrefix("10.0.0.0/8")
+
+	server := &Server{
+		trustedProxies: &StaticIPRange{
+			ranges: []netip.Prefix{localPrivatePrefix},
+		},
+		ClientIPHeaders: []string{"X-Forwarded-For"},
+	}
+
+	req := httptest.NewRequest("GET", "/", nil)
+	req.RemoteAddr = "10.0.0.1:12345"
+	req.Header.Set("X-Forwarded-For", "30.30.30.30, 45.54.45.54, 10.0.0.1")
+
+	trusted, clientIP := determineTrustedProxy(req, server)
+
+	assert.True(t, trusted)
+	assert.Equal(t, clientIP, "45.54.45.54")
```

Comment:
In reality, caddy returns `30.30.30.30` here. The validation logic in caddy is: if remoteAddress is trusted, take first value from trusted XFF header (x-forwarded-for by default, or others).

The validation logic should instead be:

if remote address is trusted, starting from the right of XFF header, return the first IP address that is untrusted.